### ISSUE TITLE
Getting -nans avg. loss while training yolov2-voc

### ISSUE THREAD

**Issue by fiorano10** [AlexeyAB/darknet#2694]
Ubuntu 16.04 with 1050Ti, while trying to retrain yolov2 with 2 classes, followed the exact steps mentioned [here](https://pjreddie.com/darknet/yolov2/). After about 1000 iterations I start getting nans for all values including avg. loss
Here is the:
[voc.data](https://drive.google.com/file/d/1YLmWfz3Z5aW-Mv9wZrmkbg68CpfQHpJg/view?usp=sharingl)
[voc.names](https://drive.google.com/file/d/1By0USLv_tu8KgaphpsD31jKmMGf_gt_3/view?usp=sharing)
[yolov2-voc.cfg](https://drive.google.com/file/d/1KeRL7poCP-6TqxJgqrMtq-kDTFYUY5M-/view?usp=sharing)
[darknet19_448.conv](https://pjreddie.com/media/files/darknet19_448.conv.23)
Initially I get this:
[5 iterations](https://drive.google.com/file/d/1Q66dzytXKiKF7Id9ikwazGRpc0xn-UsU/view?usp=sharing)
But after running it for few hours I get [1900 iterations](https://drive.google.com/file/d/1ijQma3nLcdrxRW2C1Syy0vwx3d-zS7XM/view?usp=sharing)

**Comment by AlexeyAB** [comment]
Try to train from the begining with:
```
batch=64
subdivisions=32
```

* Did you create labels for PascalVOC for only 2 classes instead of 20 classes?
* If you still get such issue, than try to train your model by using this repository https://github.com/AlexeyAB/darknet instead of https://github.com/pjreddie/darknet

**Comment by fiorano10** [comment]
I already tried `batch=64` and `subdivisions=32` and got the same result, but I'll try training from beginning.

- I generated labels for all 20 classes, I'll try generating labels for 2 classes instead.
- Which cfg files to use though? There are cfg files in multiple subdir

**Comment by fiorano10** [comment]
@AlexeyAB 
After changing `batch = 64` and `subdivisions=32` and generating labels just for 2 classes, I get random boxes after 200 iterations. But after 300 iterations I get no predictions at all.

[yolov2-voc_200.weights](https://drive.google.com/file/d/10qjLzW6t-r2WV7Fmo4S94AZhB0_QJfoX/view?usp=sharing)
[random prediction](https://drive.google.com/file/d/197xhUbERItQdNUvJptVpvEvfOiy4aECJ/view?usp=sharing)
[yolov2-voc_300.weights](https://drive.google.com/file/d/1VmamKwqFmNhDT58r0GvdwV-0KibqyM2p/view?usp=sharing)
[No predictions](https://drive.google.com/file/d/1A8JB27nFP7fGHeY9YMDfpXPRk9ualVfA/view?usp=sharing)

**Comment by AlexeyAB** [comment]
@fiorano10 Train 4000 iterations https://github.com/AlexeyAB/darknet#when-should-i-stop-training

> Usually sufficient 2000 iterations for each class(object), but not less than 4000 iterations in total.

**Comment by fiorano10** [comment]
The "Obj" and "No Obj" values descend to `~0.02`

**Comment by fiorano10** [comment]
@AlexeyAB When using `darknet/cfg/voc.data` and `darknet/cfg/yolo-v2-voc.cfg` I get really low values for "Obj" and "No Obj", should I generate new anchors even though I'm training on the PascalVOC dataset?

**Comment by AlexeyAB** [comment]
@fiorano10 What mAP do you get at 4000 iterations?

**Comment by fiorano10** [comment]
@AlexeyAB I get a mAP of 2.5% after 1000 iterations.

**Comment by fiorano10** [comment]
@AlexeyAB 
Training yolov2-voc-tiny for 2 classes, `batch=64` `subdivisions=16` `learning_rate=0.001` I'm getting a mAP of 11% after 6000 iterations

**Comment by AlexeyAB** [comment]
@fiorano10 
Do you check mAP on Validation dataset `2007_test.txt` or on Training dataset?

After 40200 iterations you should get mAP ~= 53% - 57%

**Comment by fiorano10** [comment]
Checking the mAP on the validation data set `2007_test.txt`,
`mAP(26000 iterations) = 26%`

**Comment by fiorano10** [comment]
@AlexeyAB mAP( 36000 iterations) = 26.3% still, shouldn't it be going up?

**Comment by AlexeyAB** [comment]
It should go up.
Can you share your cfg-file?
Do you train with random=1?
What command for training do you use?
What params did you use in the makefile?
And do you use for training images: `2007_train.txt 2007_val.txt 2012_train.txt 2012_val.txt > train.txt` ?

**Comment by fiorano10** [comment]
@AlexeyAB Here's the cfg file [yolov2-tiny-voc.txt](https://drive.google.com/file/d/1hrdDpUFW6GT2quZU1eRgfTjHAKz2aLVV/view?usp=sharing)
Yup, `random = 1`
command: `./darknet detector train cfg/voc.data cfg/yolov2-tiny-voc.cfg backup/yolov2-tiny-voc_last.weights -map`
`2007_train.txt 2007_val.txt 2012_train.txt 2012_val.txt > train.txt`

Note: I generated labels for only 2 classes that I wanted it to train on.

**Comment by AlexeyAB** [comment]
All is OK.

> Note: I generated labels for only 2 classes that I wanted it to train on.

Just check your dataset, that you did it correctly by using Yolo_mark: https://github.com/AlexeyAB/Yolo_mark

**Comment by fiorano10** [comment]
@AlexeyAB 
mAP (36000 iteration) went up to 28%, should it be going up faster? Or because I keep pausing it have an effect on mAP?

**Comment by fiorano10** [comment]
@AlexeyAB When training for 2 classes, do I need to generate labels for all 20 classes or just those 2? And then change the config params based on 2 classes

**Comment by AlexeyAB** [comment]
@fiorano10 You should generate labels for 2 classes

**Comment by fiorano10** [comment]
@AlexeyAB What does it use for negative classes? Does it generate labels for unlabelled images in the dataset?
